linera-core/src/value_cache.rs

```diff
@@ -184,10 +179,44 @@ impl CertificateValueCache {
     ) {
         let mut cache = self.cache.lock().await;
         for value in values {
-            let hash = value.hash();
+            let hash = (*value).hash();
             if !cache.contains(&hash) {
                 cache.push(hash, value.into_owned());
             }
         }
     }
+
+    /// Populates a [`LiteCertificate`] with its [`CertificateValue`], if it's present in
+    /// the cache.
+    pub async fn full_certificate(
+        &self,
+        certificate: LiteCertificate<'_>,
+    ) -> Result<Certificate, WorkerError> {
+        let value = self
+            .get(&certificate.value.value_hash)
+            .await
+            .ok_or(WorkerError::MissingCertificateValue)?;
+        certificate
+            .with_value(value)
+            .ok_or(WorkerError::InvalidLiteCertificate)
+    }
+}
+
+impl ValueCache<BlobId, HashedBlob> {
+    /// Inserts a [`HashedBlob`] into the cache, if it's not already present.
+    ///
+    /// The `value` is wrapped in a [`Cow`] so that it is only cloned if it needs to be
+    /// inserted in the cache.
+    ///
+    /// Returns [`true`] if the value was not already present in the cache.
+    pub async fn insert<'a>(&self, value: Cow<'a, HashedBlob>) -> bool {
+        let blob_id = (*value).id();
+        let mut cache = self.cache.lock().await;
+        if cache.contains(&blob_id) {
+            return false;
+        }
+        // Cache the blob so that clients don't have to send it again.
+        cache.push(blob_id, value.into_owned());
+        true
```

Comment:
Don't forget to promote the value if it's already in the cache:

```suggestion
        if cache.contains(&blob_id) {
            // Promote the re-inserted blob in the cache, as if it was accessed again.
            cache.promote(&blob_id);
            false
        } else {
            // Cache the blob so that clients don't have to send it again.
            cache.push(blob_id, value.into_owned());
            true
        }
```